src/spikeinterface/extractors/nwbextractors.py

```diff
@@ -166,102 +249,345 @@ def read_nwbfile(
 
     Examples
     --------
-    >>> nwbfile = read_nwbfile("data.nwb", stream_mode="ros3")
+    >>> nwbfile = read_nwbfile(file_path="data.nwb", backend="hdf5", stream_mode="ros3")
     """
-    from pynwb import NWBHDF5IO
 
     if file_path is not None and file is not None:
         raise ValueError("Provide either file_path or file, not both")
     if file_path is None and file is None:
         raise ValueError("Provide either file_path or file")
 
-    hdf5_file = _read_hdf5_file(
+    open_file = read_file_from_backend(
         file_path=file_path,
         file=file,
         stream_mode=stream_mode,
         cache=cache,
         stream_cache_path=stream_cache_path,
+        storage_options=storage_options,
     )
-    io = NWBHDF5IO(file=hdf5_file, mode="r", load_namespaces=True)
+    if backend == "hdf5":
+        from pynwb import NWBHDF5IO
+
+        io = NWBHDF5IO(file=open_file, mode="r", load_namespaces=True)
+    else:
+        from hdmf_zarr import NWBZarrIO
+
+        io = NWBZarrIO(path=open_file.store, mode="r", load_namespaces=True)
+
     nwbfile = io.read()
     return nwbfile
 
 
-class _NwbPynwbRecordingExtractor(BaseRecording):
+def get_backend_from_local_file(file_path: str | Path) -> str:
```

Comment:
This can be done with magic numbers instead of try-except:

```python
def is_hdf5_file(filename):
    # Source for magic numbers https://www.loc.gov/preservation/digital/formats/fdd/fdd000229.shtml
    # We should find a better one though
    with open(filename, 'rb') as f:
        file_signature = f.read(8)
    return file_signature == b'\x89HDF\r\n\x1a\n'
```

Reply:
Note that this will be way faster specially in the streaming case.

Reply:
Thanks!

Reply:
Note that this is only run in local mode. For streaming, the streaming mode will "know" its backend

Reply:
What do you mean? If it is fsspec how do you know if you are streaming a h5 or zarr or something else?

Reply:
`fsspec` is only available for HDF5. The `zarr` streaming mode makes use of `fsspec` too, but this is handled by the Zarr API directly

Reply:
Fsspec is not related at all to hdf5 is it? It is just a way to establis a file system in a remote location with a consistent API.

Reply:
In our usage, fsspec is used to open a remote h5 file, so we don't provide a general API to read any remote file with fsspec, but only to read remote NWB h5 files.

Reply:
Yeah, I did not generalize that function enough earlier when I made it. Ideally the file should be fetched and then either open with zarr or hdf5. But that's a matter for another PR I feel.

Reply:
I think there is no need to do that with zarr. `file = zarr.open("s3:/...")` will automatically stream using fsspec, so we shouldn't reinvent the wheel there :)